### PR TITLE
Add node builder

### PR DESCRIPTION
**Pull request by joshma**
This builder takes in language and entrypoint and finds the prjoectRoot based
on the nearest directory that contains package.json.

From there, if package-lock.json is found, it'll use npm install. Or
else, it'll use yarn.

It'll use typescript's tsc to compile to a build directory, and then
sets a Docker entrypoint running the entrypoint script out of the build
directory.

### CODE REVIEW

File: src/buildpack.ts [hunk 1]
```diff
@@ -1,31 +1,41 @@
...
+        entrypoint: string;
+      };
+    }
   | {
-    builder: "python";
-    builderConfig: {
-      entrypoint: string;
-    };
-  }
+      builder: "node-typescript";
```
Comment (colinking) [inline src/buildpack.ts:19]: Thoughts on making TS just a builder config option? 

```
{
builder: "node"
builderConfig: {
entrypoint: string
language: "typescript" | "javascript"
}
}
```

File: src/buildpack.ts [hunk 2]
```diff
@@ -60,10 +70,68 @@ export async function getDockerfile(b: Builder): Promise<string> {
...
+        workingDir = join(...pathParts);
+        break;
+      }
+      pathParts.pop();
+    }
+    if (workingDir === null) {
+      throw new Error(
+        `Could not find package.json in any directories above ${b.builderConfig.entrypoint}`
+      );
+    }
```
Comment (colinking) [inline src/buildpack.ts:90]: My last PR actually just added a helper for this

```suggestion
const workingDir = find("package.json", dirname(entrypoint))
if (!workingDir) {...}
```

File: src/buildpack.ts [hunk 3]
```diff
@@ -60,10 +70,68 @@ export async function getDockerfile(b: Builder): Promise<string> {
...
+      core.info(`Using default install command: ${installCommand}`);
+    }
+    // Produce a Dockerfile
+    const buildDir = ".airplane-build";
+    const relativeEntrypointJS = relative(
+      workingDir,
+      b.builderConfig.entrypoint
+    ).replace(/\.ts$/, ".js");
+    contents = `
+      FROM node:${NODE_VERSION}-stretch
```
Comment (colinking) [inline src/buildpack.ts:112]: Any reason for stretch vs buster?

File: src/buildpack.ts [hunk 4]
```diff
@@ -60,10 +70,68 @@ export async function getDockerfile(b: Builder): Promise<string> {
...
+      WORKDIR /airplane
+      
+      COPY ${installFiles.join(" ")} ./
+      RUN ${installCommand}
+      
+      COPY ${workingDir} ./
+      RUN echo "Cleaning ${buildDir} in case it exists" \
+          && rm -rf ${buildDir}/ \
+          && echo "Running tsc" \
+          && tsc --outDir ${buildDir}/ --rootDir .
```
Comment (colinking) [inline src/buildpack.ts:124]: Does this not need a tsconfig?

I wonder if we should just copy the entire repo over.

File: src/buildpack.ts [hunk 5]
```diff
@@ -60,10 +70,68 @@ export async function getDockerfile(b: Builder): Promise<string> {
...
+    `;
   } else if (b.builder === "python") {
-    const requirementsPath = await find("requirements.txt", dirname(b.builderConfig.entrypoint));
+    const requirementsPath = await find(
+      "requirements.txt",
+      dirname(b.builderConfig.entrypoint)
+    );
     if (!requirementsPath) {
-      throw new Error('Unable to find a requirements.txt')
+      throw new Error("Unable to find a requirements.txt");
```
Comment (colinking) [inline src/buildpack.ts:134]: Sorry, I should hook up prettier here too haha.

**Review by colinking — APPROVED** [review]

File: src/buildpack.ts (hunk 3 above)
Comment (joshma) [reply src/buildpack.ts:112]: Ah oops should be buster, the newer stable release

File: src/buildpack.ts (hunk 4 above)
Comment (joshma) [reply src/buildpack.ts]: Hm I'd expect a tsconfig.json alongside the package.json - I'm going to assume that, and write a default tsconfig.json if it doesn't exist.

**Pull request merged**